engine/access/state_stream/backend/handler.go

```diff
@@ -88,38 +102,80 @@ func (h *Handler) SubscribeExecutionData(request *executiondata.SubscribeExecuti
 
 	sub := h.api.SubscribeExecutionData(stream.Context(), startBlockID, request.GetStartBlockHeight())
 
-	for {
-		v, ok := <-sub.Channel()
-		if !ok {
-			if sub.Err() != nil {
-				return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
-			}
-			return nil
-		}
+	return subscription.HandleSubscription(sub, handleSubscribeExecutionData(stream.Send, request.GetEventEncodingVersion()))
+}
 
-		resp, ok := v.(*ExecutionDataResponse)
-		if !ok {
-			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
-		}
+// SubscribeExecutionDataFromStartBlockID handles subscription requests for
+// execution data starting at the specified block ID. The handler manages the
+// subscription and sends the subscribed information to the client via the
+// provided stream.
+//
+// Expected errors during normal operation:
+// - codes.InvalidArgument   - if request contains invalid startBlockID.
+// - codes.ResourceExhausted - if the maximum number of streams is reached.
+// - codes.Internal          - if stream got unexpected response or could not send response.
+func (h *Handler) SubscribeExecutionDataFromStartBlockID(request *executiondata.SubscribeExecutionDataFromStartBlockIDRequest, stream executiondata.ExecutionDataAPI_SubscribeExecutionDataFromStartBlockIDServer) error {
+	// check if the maximum number of streams is reached
+	if h.StreamCount.Load() >= h.MaxStreams {
+		return status.Errorf(codes.ResourceExhausted, "maximum number of streams reached")
+	}
+	h.StreamCount.Add(1)
+	defer h.StreamCount.Add(-1)
 
-		execData, err := convert.BlockExecutionDataToMessage(resp.ExecutionData)
+	startBlockID := flow.ZeroID
+	if request.GetStartBlockId() != nil {
```

Comment:
let's remove this conditional. This endpoint should require a valid start block is passed.